### ISSUE TITLE
Add an analyzer to recommend using named parameters when passing literals as arguments

### ISSUE THREAD

**Issue by DaRosenberg** [dotnet/roslyn-analyzers#3335]
We employ a guideline to always use named parameters when the meaning of an argument is not obvious from the argument itself.

For example, the intent is not very clear here, from just looking at the code:
```cs
var employees = await FetchEmployeesAsync(true);
```

If you're in an IDE, you can of course hover over the method call for a few seconds to get a tooltip, or go to the definition of the method, but it adds friction and delay when reading code. And oftentimes, code is read outside of an IDE (for example, when reviewing a PR or perusing code on GitHub).

But it's much clearer if you do:
```cs
var employees = await FetchEmployeesAsync(includeContractors: true);
```

Or:
```cs
var includeContractors = true;
var employees = await FetchEmployeesAsync(includeContractors);
```

For most cases, this could be formalized as "always use named parameters when passing literal values as arguments in a method call".

It would be nice to have an analyzer for this. Let me know if you think this is a good idea or not. I wouldn't mind taking a stab at implementing it myself, if it doesn't already exist and nobody else is already working on it.

**Comment by mavasani** [comment]
This seems related to https://github.com/dotnet/roslyn/pull/22010, which was attempted by @jcouv. I think the last comment on the PR https://github.com/dotnet/roslyn/pull/22010#issuecomment-355157341 mentions the issue with this request - the analyzer could end up being extremely demanding. So, I think we would need to implement this with following additions:

1. Ability to exclude parameter types: Have an editorconfig option to ignore specific parameter types, such as bool, which can come in lot of methods
2. Ability to exclude certain methods: Have an editorconfig option to ignore validation for specific method or property invocation, with reasonable defaults for methods/properties always excluded from this analyzer. For example, `Debug.Assert` would be a classic case that users don't care about and do not want named argument. Maybe we should ignore all methods with a single boolean parameter by default?

Tagging @Evangelink in case he is interested in this analyzer. I have always wanted to enable this analyzer for this repo itself :-)

**Comment by jcouv** [comment]
If I were to do it again today, maybe I would suppress those diagnostics (or just not enable the enforcement) on test projects altogether. Then maybe we don't need additional configuration (excluding certain methods or types).

@mavasani Let me re-open my old PR, then test on Roslyn with above rule (only enforce in product code) to see if that would be workable.

**Comment by DaRosenberg** [comment]
If the analyzer excluded certain very common cases, that'd be ideal IMO.

Besides any method with "Assert" in its name as already mentioned, also the case `ConfigureAwait(false)` comes to mind as one that is so common and well-understood that it would make sense to exclude it.

**Comment by mavasani** [comment]
@vatsalyaagrawal @jinujoseph @sharwell Can we port this issue to Roslyn? This is a code style analyzer proposal.
